MapboxNavigation/RatingControl.swift

```diff
@@ -0,0 +1,140 @@
+    }
+    
+    private func addStars() {
+        for index in 0..<starCount {
+            let button = UIButton(type: .custom)
+            button.setImage(starTemplate, for: .normal)
+            button.adjustsImageWhenHighlighted = false
+            addButtonSizeConstraints(to: button)
+            
+            let setRatingNumber = NSNumber(value: index + 1)
+            let setRatingString = NumberFormatter.localizedString(from: setRatingNumber, number: .none)
+            let localizedStarLabel = NSLocalizedString("Set %@ star rating", comment: "Accessibility Star Label")
+            button.accessibilityLabel = String.localizedStringWithFormat(localizedStarLabel, setRatingString)
+                
+            
+            button.addTarget(self, action: #selector(RatingControl.ratingButtonTapped(button:)), for: .touchUpInside)
+            
+            addArrangedSubview(button)
+            
+            stars.append(button)
+        }
+    }
+    
+    private func removeStars() {
+        stars.forEach {
+            removeArrangedSubview($0)
+            $0.removeFromSuperview()
+        }
+        stars.removeAll()
+    }
+    
+    private func updateSelectionStates() {
+        for (index, button) in stars.enumerated() {
+            
+            let selected = index < rating
+            button.tintColor = selected  ? selectedColor : normalColor
+            button.isSelected = selected
+            
+            setAccessibility(for: button, at: index)
+        }
+    }
+    
+    private func setAccessibility(for button: UIButton, at index: Int) {
+        setAccessibilityHint(for: button, at: index)
+        
+        let value: String
+        
+        switch rating {
+        case 0:
+            value = "No rating set."
+        case 1:
+            value = "1 star set."
+        default:
+            value = "\(rating) stars set."
+        }
+        
+        button.accessibilityValue = value
+    }
+    
+    private func setAccessibilityHint(for button: UIButton, at index: Int) {
+        guard rating == (index + 1) else { return } //This applies only to the zero-resettable button.
+        
+        button.accessibilityHint = NSLocalizedString("Tap to reset the rating to zero.", comment: "Rating Reset To Zero Accessability Hint")
+    }
+    
+    
+    private func addButtonSizeConstraints(to view: UIView) {
+        let width = NSLayoutConstraint(item: view, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: starSize.width)
+        let height = NSLayoutConstraint(item: view, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: starSize.height)
+        
+        NSLayoutConstraint.activate([width, height])
+    }
+    
+    @objc private func ratingButtonTapped(button sender: UIButton) throws {
+        guard let index = stars.index(of: sender) else { throw RatingControlError.tappedButtonNotInArray }
```

Comment:
I’m pretty sure the raw string value of this Error bridges to Objective-C as the NSError’s localized failure reason, in which case it could theoretically be displayed to the user. In fact, I’m surprised that this method would even be called, because the Objective-C selector should be `ratingButtonTapped:error:`, which isn’t a valid action selector. Since the error appears to be responding to a programming error entirely within the developer’s control, let’s make this an assertion instead of throwing an error.

Reply:
- [ ] do this